local-interchaintest/src/tests/osmo_liquid_pooler/osmo_liquid_pooler.rs

```diff
@@ -0,0 +1,453 @@
+use crate::helpers::constants::{
+    ACC0_ADDRESS_NEUTRON, ACC1_ADDRESS_NEUTRON, LOCAL_CODE_ID_CACHE_PATH,
+    LOCAL_CODE_ID_CACHE_PATH_OSMO, OSMOSIS_FEES, POLYTONE_PATH, VALENCE_PATH,
+};
+use cosmwasm_std::{Coin, Decimal, Uint128, Uint64};
+use covenant_utils::{op_mode::ContractOperationModeConfig, PoolPriceConfig, SingleSideLpLimits};
+use cw_utils::{Duration as CwDuration, Expiration};
+use localic_std::{errors::LocalError, modules::cosmwasm::CosmWasm, relayer::Relayer};
+use localic_utils::{
+    utils::test_context::TestContext, DEFAULT_KEY, NEUTRON_CHAIN_ID, NEUTRON_CHAIN_NAME,
+    OSMOSIS_CHAIN_NAME,
+};
+use serde_json::Value;
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+use valence_covenant_single_party_pol::msg::DEFAULT_TIMEOUT;
+use valence_osmo_liquid_pooler::msg::{PartyChainInfo, PartyDenomInfo};
+
+use log::info;
+
+/// How long we will wait until the ICA must have been created, or timed out
+const ICA_TIMEOUT: Duration = Duration::from_secs(30);
+
+macro_rules! with_poll_timeout {
+    ($body:block) => {
+        let start = Instant::now();
+
+        while Instant::now().duration_since(start) < ICA_TIMEOUT {
+            $body
+        }
+    };
+}
+
+/// Tests that the osmo liquid pooler correctly handles
+/// - Normal case
+/// - Timeouts (does not advance to proxy_created)
+/// - Timeout recovery (advances after on 2nd tick after a packet timeout)
+pub fn test_osmo_liquid_pooler(test_ctx: &mut TestContext) -> Result<(), LocalError> {
+    info!("Starting osmo liquid pooler tests...");
+
+    upload_contracts(test_ctx)?;
+
+    // Tests are still deterministic. This just allows reusing the same local-ic instance between tests
+    let label_seed = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_nanos()
+        .to_string();
+
+    let salts = (
+        format!("salt1_{label_seed}"),
+        format!("salt2_{label_seed}"),
+        format!("salt3_{label_seed}"),
+    );
+
+    // See above: 3 tests, each with a different pooler
+    let holders = (
+        make_holder(test_ctx, &salts.0, &label_seed)?,
+        make_holder(test_ctx, &salts.1, &label_seed)?,
+        make_holder(test_ctx, &salts.2, &label_seed)?,
+    );
+    let poolers = (
+        make_pooler(test_ctx, holders.0, &salts.0, &label_seed)?,
+        make_pooler(test_ctx, holders.1, &salts.1, &label_seed)?,
+        make_pooler(test_ctx, holders.2, &salts.2, &label_seed)?,
+    );
+
+    // Separate tests: ensure that the pooler can handle the happy case
+    // and a timeout (independently)
+    test_pooler_ok(test_ctx, &poolers.0)?;
+    test_pooler_timeout(test_ctx, &poolers.1)?;
+
+    // Combined test: runs test_timeout and test_ok in sequence with the same pooler
+    // to ensure that the pooler can recover from a timeout
+    test_pooler_timeout_recover(test_ctx, &poolers.2)?;
+
+    info!("Finished osmo liquid pooler tests!");
+
+    Ok(())
+}
+
+fn start_relayer(test_ctx: &mut TestContext) {
```

Comment:
we can remove these 2 functions and use the new ones you added to localic-utils right?

Reply:
Pushed the change!